src/Adapter/Swoole/EventLoop.php

```diff
@@ -0,0 +1,298 @@
+
+                    return;
+                }
+            }
+
+            // Last resolved promise resolved globalPromise
+            if (--$nbPromises === 0) {
+                $globalPromise->resolve($allResults);
+            }
+        };
+
+        foreach ($promises as $index => $promise) {
+            $this->async($waitOnePromise($index, $promise));
+        }
+
+        return $globalPromise;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function promiseForeach($traversable, callable $function): Promise
+    {
+        $promises = [];
+        foreach ($traversable as $key => $value) {
+            $promises[] = $this->async($function($value, $key));
+        }
+
+        return $this->promiseAll(...$promises);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function promiseRace(Promise ...$promises): Promise
+    {
+        if (empty($promises)) {
+            return $this->promiseFulfilled(null);
+        }
+
+        $globalPromise = $this->createPromise();
+        $isFirstPromise = true;
+
+        $wrapPromise = function (Promise $promise) use ($globalPromise, &$isFirstPromise): \Generator {
+            try {
+                $result = yield $promise;
+                if ($isFirstPromise) {
+                    $isFirstPromise = false;
+                    $globalPromise->resolve($result);
+                }
+            } catch (\Throwable $throwable) {
+                if ($isFirstPromise) {
+                    $isFirstPromise = false;
+                    $globalPromise->reject($throwable);
+                }
+            }
+        };
+
+        foreach ($promises as $promise) {
+            $this->async($wrapPromise($promise));
+        }
+
+        return $globalPromise;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function promiseFulfilled($value): Promise
+    {
+        $promise = $this->createPromise();
+        $promise->resolve($value);
+
+        return $promise;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function promiseRejected(Throwable $throwable): Promise
+    {
+        $promise = $this->createPromise();
+        $promise->reject($throwable);
+
+        return $promise;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function idle(): Promise
+    {
+        $promise = new DummyPromise();
+        Coroutine::create(function() use ($promise) {
+            $this->pushCoroutine();
+            $promise->resolve(null);
+            //Coroutine::defer(function () use ($promise) {
+            //});
+        });
+
+        return $promise;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function delay(int $milliseconds): Promise
+    {
+        $promise = new DummyPromise();
+        Coroutine::create(function() use($milliseconds, $promise) {
+            $this->pushCoroutine();
+            //Coroutine::sleep($milliseconds / 1000);
+            usleep($milliseconds * 1000);
```

Comment:
Why don't you use `Coroutine::sleep($milliseconds / 1000);`? It's exactly what we want! 😅

Reply:
Some tests are not passing (eg `testDelay` with 42 ms) as Coroutine::sleep does not accept less than 0.001 and this throws an error.
May be we should add a condition for that case.